### PR TITLE
Functioning comments with EIP-712 signatures

### LINKED ISSUE #6: Proposal Comment API

### LINKED ISSUE #182: Validate comment signature
#181 is a prerequisite. The signed message of the comment should be sent to the backend, and validated before allowing it to be posted. As a bonus, it would be nice if the comment table also held the signature to allow independent validation.

### PR DESCRIPTION
**Pull request by wbobeirne**
Closes #182. Closes #6.

### What This Does

Implements everything needed for proposal comments:

* Correct EIP-712 format signature when you try to post a comment
* Comment creation endpoint that requires auth and validates EIP-712, along with matching text
* Handling of parent comment IDs for nested comments
* Hides comment form if logged out

Some other misc improvments:

* Domain checking for EIP-712.
* Improvements to the accuracy of the `User` type with the backend `UserSchema`

At some point I think the UI here is going to need a minor facelift, but it works for now.

### How to Test

Make sure you run `flask db upgrade`. Expects you to be on version `312db8611967`.

1. Go to a proposal logged out, confirm you can't comment
2. Login, return to proposal, and confirm you can comment
3. Create a new comment, confirm it showed up
4. Create a reply, confirm it showed up
5. Refresh the page, confirm the comments loaded in correctly

### Comment EIP-712 request

screen shot 2018-11-08 at 2 30 16 pm

### Nested comments

screen shot 2018-11-08 at 2 16 04 pm

### CODE REVIEW

File: backend/grant/comment/models.py [hunk 1]
```diff
@@ -26,16 +31,21 @@ class Meta:
...
             "date_created",
             "body",
+            "replies"
         )
 
-    body = ma.Method("get_body")
 
+    body = ma.Method("get_body")
     date_created = ma.Method("get_date_created")
+    author = ma.Nested("UserSchema", exclude=["email_address"])
```
Comment (wbobeirne) [inline backend/grant/comment/models.py]: @dternyak I'd like to come up with a better strategy than ad-hoc excludes as we use this more (e.g. for proposal team), let me know if you know a better way.

**Comment by AMStrix** [comment]
Creating a proposal while signed out leaves me hanging on a 401:
screen shot 2018-11-09 at 1 50 39 pm

**Comment by AMStrix** [comment]
Should we clear the comment after successful submission?
screen shot 2018-11-09 at 1 58 18 pm

**Comment by wbobeirne** [comment]
401 error: We should make create proposal an AuthedRoute, and have it redirect to login. Or let them get to the end, and if they try to publish, tell them they need to log in.

Comment clear: Yes we should, but unfortunately this is trickier than I anticipated. For one thing, the ReactMde component we're using is an uncontrolled component, meaning it does its own internal thing, and you can only listen on changes, not update it with them. So it's a little hard to clear the component. It's also a little tricky to tell when a comment has been submitted, since it's all handled in Redux. I'll noodle on this one a little more, but it's annoyingly hard.

File: frontend/client/modules/proposals/actions.ts [hunk 2]
```diff
@@ -111,46 +112,43 @@ export function fetchProposalUpdates(proposalId: ProposalWithCrowdFund['proposal
...
+      const res = await apiPostProposalComment({
+        proposalId,
+        parentCommentId,
+        comment,
+        signedMessage: sigData.signedMessage,
+        rawTypedData: JSON.stringify(sigData.rawTypedData),
+      });
 
       // TODO: API up the comment & signed comment, handle response / failures
       // TODO: Remove console log
```
Comment (AMStrix) [inline frontend/client/modules/proposals/actions.ts]: TODO: remove todos?

File: backend/grant/comment/models.py (hunk 1 above)
Comment (dternyak) [reply backend/grant/comment/models.py]: I think the only other option would be to simply create a new "PartialUserSchema" and use that to serialize instead.

In PartialUserSchema you could filter the fields with a whitelist `fields`.

**Pull request merged**